### PR TITLE
gmc: add authN & authZ support on keycloak

### PR DESCRIPTION
**Pull request by Ruoyu-y**
## Description

Support the case that using the open source project Keycloak as the identity management tool and provide authentication materials (JWT tokens).

## Issues

List the issue or RFC link this PR is working on. If there is no such link, please mark it as `n/a`.

## Type of change

- [ * ] New feature (non-breaking change which adds new functionality)

* add gateway, authN and authZ yaml for keycloak option
* add documentation for keycloak option

## Dependencies

Keycloak
Istio

### CODE REVIEW

**Comment by Ruoyu-y** [comment]
@irisdingbj @zhlsunshine Here's the PR for AuthN & AuthZ with Keycloak. Please take a look and help with the review.

File: microservices-connector/config/authN-authZ/README.md [hunk 1]
```diff
@@ -79,5 +80,77 @@ kubectl exec -it -n chatqa $CLIENT_POD -- curl -X POST $accessUrl -d '{"text":"W
...
+
+Access the Keycloak admin console through the `KEYCLOAK_ADDR` to configure the users. Use the username and password specified in the yaml file to login. Configuration steps look like this:
+
+1. Create a new realm named `istio` within Keycloak.
+
+2. Create a new client called `istio` with default configurations.
+
+3. From the left pane select the Realm roles and create a new role name as `user` and another new role as `viewer`.
+
+4. Create a new user name as `users` and another user as `userb`. Set passwords for both users (set 'Temporary' to 'Off'). Select Role mapping on the top, assign the `user` role to `users` and assign the `viewer` role to `userb`.
```
Comment (irisdingbj) [inline microservices-connector/config/authN-authZ/README.md:113]: Can we give another more meaningfull user name add some explanations for these two users?  and explain the scenario below.

Comment (Ruoyu-y) [reply microservices-connector/config/authN-authZ/README.md]: I changed the username to 'mary' and 'bob' in the case and add a simple explanation under the section "Perform authentication and authorization via JWT tokens generated by Keycloak"

File: microservices-connector/config/authN-authZ/README.md [hunk 2]
```diff
@@ -75,9 +76,86 @@ export accessUrl=$(kubectl get gmc -n chatqa -o jsonpath="{.items[?(@.metadata.n
...
+envsubst < $(pwd)/config/authN-authZ/chatQnA_authN_keycloak.yaml | kubectl -n chatqa apply -f -
+envsubst < $(pwd)/config/authN-authZ/chatQnA_authZ_keycloak.yaml | kubectl -n chatqa apply -f -
+```
+
+**Validate authentication and authorization with different JWT tokens**
+
+Fetch ID tokens for the two different users. **Note:** Remember to replace the `password` in the curl url and turn off the all the 'Required actions' under the 'Authentication' section.
+
+```bash
+# get id token for users. Please replace the password inside the url.
```
Comment (irisdingbj) [inline microservices-connector/config/authN-authZ/README.md:140]: s/for users/for Mary

File: microservices-connector/config/authN-authZ/README.md [hunk 3]
```diff
@@ -75,9 +76,86 @@ export accessUrl=$(kubectl get gmc -n chatqa -o jsonpath="{.items[?(@.metadata.n
...
+export TOKENA=$(curl -X POST 'http://${KEYCLOAK_ADDR}/realms/istio/protocol/openid-connect/token' -H 'Content-Type: application/x-www-form-urlencoded' --data-urlencode 'grant_type=password' --data-urlencode 'client_id=istio' --data-urlencode 'username=mary' -d 'password=${PASSWORD}' -d 'scope=openid' -d 'response_type=id_token' | jq -r .id_token)
+
+# get id token for userb. Please replace the password inside the url.
```
Comment (irisdingbj) [inline microservices-connector/config/authN-authZ/README.md:143]: s/ for userb/ for bob

File: microservices-connector/config/authN-authZ/README.md (hunk 2 above)
Comment (Ruoyu-y) [reply microservices-connector/config/authN-authZ/README.md]: Done

File: microservices-connector/config/authN-authZ/README.md (hunk 3 above)
Comment (Ruoyu-y) [reply microservices-connector/config/authN-authZ/README.md]: Done

**Review by irisdingbj — APPROVED** [review]

**Review by zhlsunshine — APPROVED** [review]
LGTM

**Pull request merged**